clients/drcachesim/scheduler/scheduler_impl.cpp

```diff
@@ -2279,7 +2324,27 @@ scheduler_impl_tmpl_t<RecordType, ReaderType>::get_next_trace_pc(output_ordinal_
         inputs_[index].cur_queue_record.next_trace_pc_valid) {
         return inputs_[index].cur_queue_record.next_trace_pc;
     }
-    return inputs_[index].reader.get()->get_next_trace_pc();
+    // If the queue's front has a record that has a pc or it has a snapshotted value
+    // of the next pc, use that.
+    uint64_t pc;
+    if (!inputs_[index].queue.empty()) {
+        if (record_type_has_pc(inputs_[index].queue.front().record, pc)) {
+            return pc;
+        }
+        if (inputs_[index].queue.front().next_trace_pc_valid) {
```

Comment:
Does this also cover the has_pc case and we could removes lines 2331-2333?

Reply:
No the has_pc case above is different.

Consider: If the record at the front of the queue is an instr itself, its next_trace_pc field would hold the pc of the yet-next instr. But for us (we're trying to determine the next trace pc at a time when that front-of-the-queue instr is also not yet returned), we want to return the front-of-the-queue-instr's pc.